# Build a regression model.

In [1]:
# Import the packages we will need
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [5]:
# Read the dataset into the notebook
df = pd.read_csv('final_df.csv', index_col=False)
df = df.drop('Unnamed: 0', axis = 1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 224 entries, 0 to 223
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   station_name                   224 non-null    object 
 1   station_id                     224 non-null    object 
 2   station_address                224 non-null    object 
 3   empty_slots                    224 non-null    int64  
 4   free_bikes                     224 non-null    int64  
 5   latitude                       224 non-null    float64
 6   longitude                      224 non-null    float64
 7   num_nearby_bars                224 non-null    int64  
 8   avg_pop_bars                   224 non-null    float64
 9   avg_rat_bars                   224 non-null    float64
 10  num_nearby_lodging             224 non-null    int64  
 11  avg_pop_lodging                224 non-null    float64
 12  avg_rat_lodging                224 non-null    flo

### Initial Model

The first model will be a multivariate analysis of the entire dataset without any scaling applied to the features

In [15]:
# Create the first features and target dataframes for the analysis
features = df.columns[7:19]
target = df.columns[-1]
X = df[features]
y = df[target]

In [16]:
X.head()

num_nearby_bars  avg_pop_bars  avg_rat_bars  num_nearby_lodging  \
0               18      0.849992      8.416667                   6   
1               46      0.849432      8.266667                  15   
2               50      0.972846      7.689796                  50   
3               33      0.905051      7.600000                  30   
4                6      0.952832      7.400000                  12   

   avg_pop_lodging  avg_rat_lodging  num_nearby_trans_hubs  \
0         0.943272         9.100000                     16   
1         0.505910         7.850000                     11   
2         0.834985         7.583333                     50   
3         0.775464         6.450000                     32   
4         0.833727         0.000000                     25   

   avg_pop_trans_hubs  avg_rating_trans_hubs  average_yelp_rating_bar  \
0            0.813149               0.000000                     4.20   
1            0.923558               0.000000                     4.17   
2            0.709939               7.083333                     4.06   
3            0.687911               7.220000                     4.15   
4            0.735822               6.500000                     4.18   

   average_yelp_rating_hotels  average_yelp_rating_transport  
0                        3.57                           4.70  
1                        3.61                           4.77  
2                        3.50                           4.72  
3                        3.61                           4.66  
4                        3.64                           4.80

In [17]:
y.head()

0    20
1    18
2    31
3    17
4    15
Name: total_bike_slots, dtype: int64

In [18]:
X = sm.add_constant(X) # add a constant
lin_reg = sm.OLS(y,X)
X

const  num_nearby_bars  avg_pop_bars  avg_rat_bars  num_nearby_lodging  \
0      1.0               18      0.849992      8.416667                   6   
1      1.0               46      0.849432      8.266667                  15   
2      1.0               50      0.972846      7.689796                  50   
3      1.0               33      0.905051      7.600000                  30   
4      1.0                6      0.952832      7.400000                  12   
..     ...              ...           ...           ...                 ...   
219    1.0               31      0.859253      7.725000                  17   
220    1.0               27      0.863153      7.684615                  16   
221    1.0               50      0.970269      7.700000                  50   
222    1.0                8      0.915316      7.475000                   4   
223    1.0               36      0.857152      7.500000                  24   

     avg_pop_lodging  avg_rat_lodging  num_nearby_trans_hubs  \
0           0.943272         9.100000                     16   
1           0.505910         7.850000                     11   
2           0.834985         7.583333                     50   
3           0.775464         6.450000                     32   
4           0.833727         0.000000                     25   
..               ...              ...                    ...   
219         0.917085         6.650000                     16   
220         0.918970         6.650000                     16   
221         0.763539         7.530000                     50   
222         0.000000         0.000000                      4   
223         0.780557         6.833333                      6   

     avg_pop_trans_hubs  avg_rating_trans_hubs  average_yelp_rating_bar  \
0              0.813149               0.000000                     4.20   
1              0.923558               0.000000                     4.17   
2              0.709939               7.083333                     4.06   
3              0.687911               7.220000                     4.15   
4              0.735822               6.500000                     4.18   
..                  ...                    ...                      ...   
219            0.728487               5.800000                     4.18   
220            0.706834               5.800000                     4.18   
221            0.718122               7.083333                     4.07   
222            0.855221               0.000000                     4.19   
223            0.531032               0.000000                     4.26   

     average_yelp_rating_hotels  average_yelp_rating_transport  
0                          3.57                           4.70  
1                          3.61                           4.77  
2                          3.50                           4.72  
3                          3.61                           4.66  
4                          3.64                           4.80  
..                          ...                            ...  
219                        3.84                           4.80  
220                        3.82                           4.80  
221                        3.50                           4.73  
222                        3.86                           4.79  
223                        3.80                           4.80  

[224 rows x 13 columns]

## Provide model output and an interpretation of the results. 

In [25]:
# fit the model
model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:       total_bike_slots   R-squared:                       0.266
Model:                            OLS   Adj. R-squared:                  0.224
Method:                 Least Squares   F-statistic:                     6.373
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           1.29e-09
Time:                        16:37:39   Log-Likelihood:                -662.17
No. Observations:                 224   AIC:                             1350.
Df Residuals:                     211   BIC:                             1395.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
const         

#### Initial Model Interpretation

1. The adjusted R-squared of this model is quite low, 0.224, which translates to only being able to explain 22% of the data variations.
2. All features except for the YELP variables have a p-value above the 0.05 threshold.

### Second Model - Multivariate Regression for YELP Variables only

The second model will use only YELP variables which were shown above to be statistically significant

In [32]:
# Create the second features and target dataframes for the analysis of YELP variables
features_YELP = df.columns[16:19]
target_YELP = df.columns[-1]
X_YELP = df[features_YELP]
y_YELP = df[target_YELP]

In [33]:
# Add a constant and initiate the linear regression model
X_YELP = sm.add_constant(X_YELP)
lin_reg_YELP = sm.OLS(y_YELP,X_YELP)
X_YELP

const  average_yelp_rating_bar  average_yelp_rating_hotels  \
0      1.0                     4.20                        3.57   
1      1.0                     4.17                        3.61   
2      1.0                     4.06                        3.50   
3      1.0                     4.15                        3.61   
4      1.0                     4.18                        3.64   
..     ...                      ...                         ...   
219    1.0                     4.18                        3.84   
220    1.0                     4.18                        3.82   
221    1.0                     4.07                        3.50   
222    1.0                     4.19                        3.86   
223    1.0                     4.26                        3.80   

     average_yelp_rating_transport  
0                             4.70  
1                             4.77  
2                             4.72  
3                             4.66  
4                             4.80  
..                             ...  
219                           4.80  
220                           4.80  
221                           4.73  
222                           4.79  
223                           4.80  

[224 rows x 4 columns]

In [34]:
# fit the model
model_YELP = lin_reg_YELP.fit()
print_model_YELP = model_YELP.summary()
print(print_model_YELP)

                            OLS Regression Results                            
Dep. Variable:       total_bike_slots   R-squared:                       0.257
Model:                            OLS   Adj. R-squared:                  0.247
Method:                 Least Squares   F-statistic:                     25.40
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           3.76e-14
Time:                        16:39:26   Log-Likelihood:                -663.49
No. Observations:                 224   AIC:                             1335.
Df Residuals:                     220   BIC:                             1349.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
const         

### YELP Model Interpretation

1. The adjusted R-Squared value increased slightly from 0.224 to 0.247 which confirms that the YELP variables were primary responsible for the accuracy of the initial model.

### Third Model - Normalized YELP Model

The YELP rating variables are on a scale from 0 - 5. I will create a model that normalized these values, then runs the model again to compare against the second model.

#### Recall the histograms of the final dataset

In [20]:
import plotly.express as px

In [36]:
YELP_columns = df.columns[16:19]
for feature in YELP_columns:
    fig = px.histogram(df, x = feature)
    fig.show()

### Observations on Feature Columns

1. The average yelp rating variables for bars and transportation hubs appear to be normal. Perform check.

In [37]:
# check to see if the bars and rating features pass or fail the Shapiro-Wilk test for normality

from scipy import stats

for feature in YELP_columns:
    stat, p = stats.shapiro(df[feature])
    print('The p-value of', feature, 'is:', p)

The p-value of average_yelp_rating_bar is: 9.848700699421897e-08
The p-value of average_yelp_rating_hotels is: 3.1268014026863966e-06
The p-value of average_yelp_rating_transport is: 1.8811426500303695e-15


#### All YELP features have a p-value < 0.05 which means this data is not normally distributed

### Normalizing YELP Data

In [40]:
df_to_normalize = df.copy()

In [44]:
# normalize the below columns
normalized_YELP_columns = ['average_yelp_rating_bar','average_yelp_rating_hotels','average_yelp_rating_transport']

for feature in normalized_YELP_columns:
    normalized_YELP_feature = (df_to_normalize[feature] - df_to_normalize[feature].min()) / (df_to_normalize[feature].max() - df_to_normalize[feature].min())
    df_to_normalize[feature] = normalized_YELP_feature

df_to_normalize.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 224 entries, 0 to 223
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   station_name                   224 non-null    object 
 1   station_id                     224 non-null    object 
 2   station_address                224 non-null    object 
 3   empty_slots                    224 non-null    int64  
 4   free_bikes                     224 non-null    int64  
 5   latitude                       224 non-null    float64
 6   longitude                      224 non-null    float64
 7   num_nearby_bars                224 non-null    int64  
 8   avg_pop_bars                   224 non-null    float64
 9   avg_rat_bars                   224 non-null    float64
 10  num_nearby_lodging             224 non-null    int64  
 11  avg_pop_lodging                224 non-null    float64
 12  avg_rat_lodging                224 non-null    flo

### Histograms of Normalized YELP features

In [48]:
normalized_YELP_columns = df_to_normalize.columns[16:19]
for feature in normalized_YELP_columns:
    fig = px.histogram(df_to_normalize, x = feature)
    fig.show()

### Convert to left skew and transform with log function

In [50]:
# transform the average yelp rating bars and transportation columns
transformed_YELP_columns = ['average_yelp_rating_bar','average_yelp_rating_hotels', 'average_yelp_rating_transport']

for feature in transformed_YELP_columns:
    # subtract each value from the max to convert to left skew
    left_skew_YELP_feature = df_to_normalize[feature].max() - df_to_normalize[feature]
    # apply log function to value
    log_left_skew_YELP_feature = np.log(left_skew_YELP_feature.max() - left_skew_YELP_feature + 1)
    df_to_normalize[feature] = log_left_skew_YELP_feature

### Histograms of Normalized, Transformed Features

In [54]:
normalized_YELP_columns = df_to_normalize.columns[16:19]
for feature in normalized_YELP_columns:
    fig = px.histogram(df_to_normalize, x = feature)
    fig.show()

In [51]:
# Create the normalized YELP features and target dataframes for the analysis of the normalized YELP features
features_YELP_norm = df_to_normalize.columns[16:19]
target_YELP_norm = df_to_normalize.columns[-1]
X_YELP_norm = df_to_normalize[features_YELP_norm]
y_YELP_norm = df_to_normalize[target_YELP_norm]

In [52]:
# add a constant and instantiate the model
X_YELP_norm = sm.add_constant(X_YELP_norm)
lin_reg_YELP_norm = sm.OLS(y_YELP_norm,X_YELP_norm)
X_YELP_norm

const  average_yelp_rating_bar  average_yelp_rating_hotels  \
0      1.0                 0.552790                    0.160343   
1      1.0                 0.510826                    0.231802   
2      1.0                 0.339868                    0.021506   
3      1.0                 0.481838                    0.231802   
4      1.0                 0.525010                    0.282232   
..     ...                      ...                         ...   
219    1.0                 0.525010                    0.565808   
220    1.0                 0.525010                    0.540806   
221    1.0                 0.356675                    0.021506   
222    1.0                 0.538997                    0.590199   
223    1.0                 0.631778                    0.515164   

     average_yelp_rating_transport  
0                         0.488847  
1                         0.578078  
2                         0.515164  
3                         0.434038  
4                         0.614010  
..                             ...  
219                       0.614010  
220                       0.614010  
221                       0.528067  
222                       0.602175  
223                       0.614010  

[224 rows x 4 columns]

In [53]:
# fit the model
model_YELP_norm = lin_reg_YELP_norm.fit()
print_model_YELP_norm = model_YELP_norm.summary()
print(print_model_YELP_norm)

                            OLS Regression Results                            
Dep. Variable:       total_bike_slots   R-squared:                       0.261
Model:                            OLS   Adj. R-squared:                  0.251
Method:                 Least Squares   F-statistic:                     25.88
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           2.23e-14
Time:                        17:30:59   Log-Likelihood:                -662.95
No. Observations:                 224   AIC:                             1334.
Df Residuals:                     220   BIC:                             1348.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
const         

### Normalized, Transformed Model Interpretation

The R-Squared vale of this model has increased again from 0.247 to 0.251 and also halved the p-value of the average yelp rating hotel feature from 0.033 to 0.016.

# Stretch

## How can you turn the regression model into a classification model?

This analysis could be turned into a classification model by setting the target variable of bike slots at each station to be categorical of above or below a certain number of bike slots.